ferveo/src/refresh.rs

```diff
@@ -0,0 +1,384 @@
+            .collect::<Vec<_>>();
+
+        // Finally, let's recreate the shared private key from some original shares and the recovered one
+        domain_points.push(x_r);
+        private_shares.push(new_private_key_share);
+        let start_from = shares_num - threshold;
+        let new_shared_private_key = recover_share_from_updated_private_shares(
+            &ScalarField::zero(),
+            &domain_points[start_from..],
+            &private_shares[start_from..],
+        );
+
+        assert_eq!(
+            shared_private_key,
+            new_shared_private_key.private_key_share
+        );
+    }
+
+    /// Ñ parties (where t <= Ñ <= N) jointly execute a "share refresh" algorithm.
+    /// The output is M new shares (with M <= Ñ), with each of the M new shares substituting the
+    /// original share (i.e., the original share is deleted).
+    #[test]
+    fn tdec_simple_variant_share_refreshing() {
+        let rng = &mut test_rng();
+        let shares_num = 16;
+        let threshold = shares_num * 2 / 3;
+
+        let (_, shared_private_key, contexts) =
+            setup_simple::<E>(threshold, shares_num, rng);
+
+        let domain_points = &contexts[0]
+            .public_decryption_contexts
+            .iter()
+            .map(|ctxt| ctxt.domain)
+            .collect::<Vec<_>>();
+        let h = contexts[0].public_decryption_contexts[0].h;
+
+        // Each participant prepares an update for each other participant:
+        let share_updates = contexts
+            .iter()
+            .map(|p| {
+                let deltas_i = prepare_share_updates_for_refresh::<E>(
+                    &domain_points,
+                    &h,
+                    threshold,
+                    rng,
+                );
+                (p.index, deltas_i)
+            })
+            .collect::<HashMap<_, _>>();
+
+        // Participants "refresh" their shares with the updates from each other:
+        let refreshed_shares: Vec<_> = contexts
+            .iter()
+            .map(|p| {
+                // Current participant receives updates from other participants
+                let updates_for_participant: Vec<_> = share_updates
+                    .values()
+                    .map(|updates| *updates.get(p.index).unwrap())
+                    .collect();
+
+                // And updates their share
+                apply_updates_to_private_share::<E>(
+                    &p.private_key_share,
+                    &updates_for_participant,
+                )
+            })
+            .collect();
+
+        // Finally, let's recreate the shared private key from the refreshed shares
+        let new_shared_private_key = recover_share_from_updated_private_shares(
+            &ScalarField::zero(),
+            &domain_points[..threshold],
+            &refreshed_shares[..threshold],
+        );
+
+        assert_eq!(
+            shared_private_key,
+            new_shared_private_key.private_key_share
+        );
```

Comment:
We could test the red path here (not enough shares) similarly as we do in `tdec_simple_variant_share_recovery_at_selected_point`